File: Easy/0141_linked-list-cycle/linked-list-cycle.go
package main

import (
	"fmt"
)

type ListNode struct {
	Val int
	Next *ListNode
}

func hasCycle(head *ListNode) bool {
	if head == nil || head.Next == nil {
		return false
	}
	slow := head
	fast := head.Next
	for fast != nil && fast.Next != nil {
		if slow == fast {
			return true
		}
		slow = slow.Next
		fast = fast.Next.Next
	}
	return false
}

func createCycleLinkedList(values []int, pos int) *ListNode {
	if len(values) == 0 {
		return nil
	}

	head := &ListNode{Val: values[0]}
	current := head
	nodes := []*ListNode{head}
	for i := 1; i < len(values); i++ {
		node := &ListNode{Val: values[i]}
		current.Next = node
		current = node
		nodes = append(nodes, node)
	}
	if pos != -1 {
		current.Next = nodes[pos]
	}
	return head
}

type testCase struct {
	nums []int
	pos int
}

func main() {
	tests := []testCase{
		{[]int{3, 2, 0, -4}, 1},
		{[]int{1,2}, 0},
		{[]int{1}, -1},
	}
	
	for _, tc := range tests {
		head := createCycleLinkedList(tc.nums, tc.pos)
		fmt.Printf("---\nList: %v pos: %d\n", tc.nums, tc.pos)
		fmt.Printf("hasCycle: %t\n", hasCycle(head))
	}
}

File: Easy/0141_linked-list-cycle/linked-list-cycle.py
#!/bin/python3

from typing import List, Optional
from dataclasses import dataclass

class ListNode:
    def __init__(self, x):
        self.val = x
        self.next = None

class Solution:
    def hasCycle(self, head: Optional[ListNode]) -> bool:
        if not head or not head.next:
            return False
        slow = head
        fast = head.next
        while fast and fast.next:
            if slow == fast:
                return True
            slow = slow.next
            fast = fast.next.next
        return False

def createCycleLinkedList(values: List[int], pos: int):
    if not values:
        return None
    head = ListNode(values[0])
    current = head
    nodes = [head]
    for val in values[1:]:
        node = ListNode(val)
        current.next = node
        current = node
        nodes.append(node)
    if pos != -1:
        current.next = nodes[pos]
    return head

@dataclass
class testCase:
    nums: List[int]
    pos: int

def main():
    tests: List[testCase] = [
        testCase([3, 2, 0, -4], 1),
        testCase([1,2], 0),
        testCase([1], -1),
    ]

    for test in tests:
        nums = test.nums
        pos = test.pos
        print(f"---\nList: {nums} pos: {pos}")
        head = createCycleLinkedList(nums, pos)
        print(f"hasCycle: {Solution().hasCycle(head)}")

if __name__ == "__main__":
    main()
